pkg/cmd/roachtest/tests/mvcc_gc.go

```diff
@@ -0,0 +1,443 @@
+
+import (
+	"context"
+	gosql "database/sql"
+	"fmt"
+	"math"
+	"math/rand"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+)
+
+func registerMVCCGC(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:    "mvcc_gc",
+		Owner:   registry.OwnerKV,
+		Timeout: 15 * time.Minute,
+		Cluster: r.MakeClusterSpec(3),
+		Run:     runMVCCGC,
+	})
+}
+
+/*
+Test verifies that GC respects range tombstones and garbage collects range
+tombstones and that mvcc stats are correctly updated.
+
+Test sequence:
+- create kv test data schema
+- pre-split with 3 min expiration to allow merging to proceed in parallel after
+  initial data is loaded
+- load test data using KV workload
+- repeat 3 times (while ranges are being gradually merged back)
+  - delete all data in kv table using range tombstones
+    (deletion is done using cli commands that sends kv requests to cluster)
+  - sets low GC ttl to make all data eligible for deletion after 2 min
+  - write more data over the existing one
+  - verify that all garbage and range keys were removed by GC within 5 min
+*/
+
+func runMVCCGC(ctx context.Context, t test.Test, c cluster.Cluster) {
+	rng, _ := randutil.NewTestRand()
+
+	// How many times test repeats generate/cleanup cycles after initial one.
+	const cleanupRuns = 3
+
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	s := install.MakeClusterSettings()
+	s.Env = append(s.Env, "COCKROACH_SCAN_INTERVAL=30s")
+	c.Start(ctx, t.L(), option.DefaultStartOpts(), s)
+
+	conn := c.Conn(ctx, t.L(), 1)
+	defer conn.Close()
+
+	execSQLOrFail := func(statement string, args ...interface{}) {
+		if _, err := conn.ExecContext(ctx, statement, args...); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	setClusterSetting := func(name string, value interface{}) {
+		execSQLOrFail(fmt.Sprintf(`SET CLUSTER SETTING %s = $1`, name), value)
+	}
+
+	setClusterSetting("storage.mvcc.range_tombstones.enabled", true)
+	setClusterSetting("kv.protectedts.poll_interval", "5s")
+	setClusterSetting("kv.mvcc_gc.queue_interval", "0s")
+
+	m := c.NewMonitor(ctx)
+	m.Go(func(ctx context.Context) error {
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "init", "kv", "--cycle-length", "10000")
+
+		t.L().Printf("finished init, doing force split ranges")
+		splitKvIntoRanges(t, conn, 100, 180*time.Second, "kv", "kv")
```

Comment:
nit: can we use `workload init kv --splits 100` instead?

Reply:
We can't in this particular case because 1) it uses split that sets expiration to far in the future preventing any merges 2) split is broken when cycle-length is defined (there's some trickiness as well since if split respects cycle-length, then sharding should be different when we rerun with different cycle).